src/dynamics/solver/joint_constraint/prismatic_velocity_constraint.rs

```diff
@@ -141,13 +141,45 @@ impl PrismaticVelocityConstraint {
         #[cfg(feature = "dim3")]
         let inv_lhs = Cholesky::new_unchecked(lhs).inverse();
 
-        let lin_rhs = basis1.tr_mul(&(anchor_linvel2 - anchor_linvel1));
-        let ang_rhs = rb2.angvel - rb1.angvel;
+        let linvel_err = basis1.tr_mul(&(anchor_linvel2 - anchor_linvel1));
+        let angvel_err = rb2.angvel - rb1.angvel;
 
         #[cfg(feature = "dim2")]
-        let rhs = Vector2::new(lin_rhs.x, ang_rhs);
+        let mut rhs = Vector2::new(linvel_err.x, angvel_err) * params.velocity_solve_fraction;
         #[cfg(feature = "dim3")]
-        let rhs = Vector5::new(lin_rhs.x, lin_rhs.y, ang_rhs.x, ang_rhs.y, ang_rhs.z);
+        let mut rhs = Vector5::new(
+            linvel_err.x,
+            linvel_err.y,
+            angvel_err.x,
+            angvel_err.y,
+            angvel_err.z,
+        ) * params.velocity_solve_fraction;
+
+        let velocity_based_erp_inv_dt = params.velocity_based_erp_inv_dt();
+        if velocity_based_erp_inv_dt != 0.0 {
+            let dpos = anchor2 - anchor1;
+            let limit_err = dpos.dot(&axis1);
+            let mut linear_err = dpos - *axis1 * limit_err;
+
+            let frame1 = rb1.position * cparams.local_frame1();
+            let frame2 = rb2.position * cparams.local_frame2();
+            let ang_err = frame2.rotation * frame1.rotation.inverse();
+
+            let (min_limit, max_limit) = (cparams.limits[0], cparams.limits[1]);
+            linear_err +=
+                *axis1 * ((limit_err - max_limit).max(0.0) - (min_limit - limit_err).max(0.0));
+
+            #[cfg(feature = "dim2")]
+            {
+                rhs += Vector2::new(linear_err.x, ang_err.angle()) * velocity_based_erp_inv_dt;
+            }
+            #[cfg(feature = "dim3")]
+            {
+                let ang_err = ang_err.scaled_axis();
+                rhs += Vector5::new(linear_err.x, linear_err.y, ang_err.x, ang_err.y, ang_err.z)
```

Comment:
I just realized we need `rhs` to be a `Vector6` here for limits to work

Reply:
Another solution is to implement prismatic limits as a separate 1dof linear limit constraint

Reply:
Since the limit is an unilateral constraint, it is preferable to model it as a separate 1D constraint. Otherwise it won't interact well with the 5 other bilateral constraints. In particular, using a `Vector6` implies that the inverse Delassus matrix (`inv_lhs`) becomes 6x6 that either:
- Couples the joint limit constraint with the others, which is invalid if the limit isn't violated.
- Or not couple them (by setting all the off-diagonal elements of the 6-th row and column of the Delassus matrix), in which case the formulation is correct but this is equivalent to solve a 1D limit constraint and a 5D hard bilateral constraint in a Jacobi fashion (instead of a Gauss-Seidel) so this would probably reduce convergence rate.

Reply:
I guess I should just use the existing `PrismaticVelocityGroundConstraint::limits_rhs`

Reply:
> I guess I should just use the existing PrismaticVelocityGroundConstraint::limits_rhs

Yes, this should do the trick.

Reply:
I'll take this opportunity to also change `limits_forcedirs` to be just one vector, as the direction of forces on the two bodies should be opposite and equal

Reply:
> I'll take this opportunity to also change limits_forcedirs to be just one vector, as the direction of forces on the two bodies should be opposite and equal

If possible it would be best to keep that change to another PR.

The reason why the `limits_forcedirs` is two vectors is to make the prismatic joint reduction of DoFs part of the constraint Jacobian so that the limit constraint does not increase the error of the other constraints whenever there is numerical drift making the axes not aligned. Though it's a bit over-simplified right now: the correct modeling is probably to project `axis1` on the kernel of the basis of the locked DoFs of the second prismatic joint to get something like this: `let axis2 = -(limit_forcedir1 - basis2 * basis2.transpose() * limit_forcedir1)`.

Reply:
prismatic narrow fixed